stats/opencensus/opencensus.go

```diff
@@ -78,15 +83,89 @@ func ServerOption(to TraceOptions) grpc.ServerOption {
 }
 
 // unaryInterceptor handles per RPC context management. It also handles per RPC
-// tracing and stats.
+// tracing and stats, and records the latency for the full RPC call.
 func unaryInterceptor(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
-	return invoker(ctx, method, req, reply, cc, opts...)
+	startTime := time.Now()
+	err := invoker(ctx, method, req, reply, cc, opts...)
+	callLatency := float64(time.Since(startTime)) / float64(time.Millisecond)
+
+	var st string
+	if err != nil {
+		s := status.Convert(err)
+		st = canonicalString(s.Code())
+	} else {
+		st = "OK"
+	}
+
+	ocstats.RecordWithOptions(ctx,
+		ocstats.WithTags(
+			tag.Upsert(keyClientMethod, removeLeadingSlash(method)),
+			tag.Upsert(keyClientStatus, st),
+		),
+		ocstats.WithMeasurements(
+			clientAPILatency.M(callLatency),
+		),
+	)
+
+	return err
+}
+
+// wrappedStream wraps a grpc.ClientStream to intercept the RecvMsg call to take
+// latency measurements on RPC Completion.
+type wrappedStream struct {
+	grpc.ClientStream
+	// The timestamp of when this stream started.
+	startTime time.Time
+	// the method for this stream.
+	method string
+}
+
+func newWrappedStream(s grpc.ClientStream) grpc.ClientStream {
+	return &wrappedStream{ClientStream: s}
+}
+
+func (ws *wrappedStream) RecvMsg(m interface{}) error {
+	err := ws.ClientStream.RecvMsg(m)
+	if err == nil {
+		// RPC isn't over yet, no need to take measurement.
+		return err
+	}
+	// RPC completed, record measurement for full call latency.
+	callLatency := float64(time.Since(ws.startTime)) / float64(time.Millisecond)
+	var st string
+	if err == io.EOF {
+		st = "OK"
+	} else {
+		s := status.Convert(err)
+		st = canonicalString(s.Code())
+	}
+
+	ocstats.RecordWithOptions(context.Background(),
```

Comment:
What happens if the user calls another method on the stream now?  It's possible we can record the same metric more than once?

Maybe we should put this inside a `Once`, and make it a method on `wrappedStream`?

Reply:
Discussed offline; switched to callback in call options that gets called on clientStream.finish(). There's a bool in that function which gates logic in that function for having the logic run once: https://github.com/grpc/grpc-go/blob/master/stream.go#L973, so putting it right after can guarantee that this only runs once.